lib/stub_ui/lib/src/ui/text.dart

```diff
@@ -1011,6 +1011,185 @@ enum BoxWidthStyle {
+  /// metrics directly reflects the intended signage of the value. For example,
+  /// the y-coordinate of the top edge of the line is 'baseline - ascent`.
+  final double ascent;
+
+  /// The drop from the [baseline] as calculated from the font and style for this line.
+  ///
+  /// The y-coordinate of the bottom edge of the line is 'baseline + descent`.
+  final double descent;
+
+  /// The rise from the [baseline] as calculated from the font and style for this line
+  /// ignoring the text scale factor.
+  ///
+  /// The [unscaledAscent] is provided as a positive value, even though it is typically
+  /// defined in fonts as negative. This is to ensure the signage of operations with
+  /// these metrics directly reflects the intended signage of the value. For example,
+  /// the y-coordinate of the top edge of the line is 'baseline - ascent`.
+  final double unscaledAscent;
+
+  /// Total height of the line from the top edge to the bottom edge.
+  ///
+  /// This is equivalent to `ascent + descent`
+  final double height;
+
+  /// Width of the line.
+  final double width;
+
+  /// The x-coordinate of left edge of the line.
+  ///
+  /// The right edge can be obtained with `left + width`.
+  final double left;
+
+  /// The y position of the baseline for this line from the top of the paragraph.
+  final double baseline;
+
+  /// Zero indexed line number.
+  ///
+  /// For example, the first line is line 0, second line is line 1.
+  final int lineNumber;
+
+  /// Mapping between text index ranges and the RunMetrics associated with
+  /// them.
+  ///
+  /// Each entry will be keyed under the ending index the run applies to. Each
+  /// [RunMetrics] entry corresponds to the text covered by the range of
+  /// `[previous key, current key)`. For a given index `i`, the [RunMetrics]
+  /// that applies to it is the largest key such that `i < key`. For the first
+  /// key, the [startIndex] acts as the previous key.
+  ///
+  /// The metrics here are pre-layout values and are the base font metrics we
+  /// compute the rest of the paragraph from. These metrics do not account for
+  /// modifiers such as [StrutStyle] or `textScaleFactor`.
+  final Map<int, RunMetrics> runMetrics;
+}
+
+/// Data class that tracks the raw font metrics for a particular text buffer index.
+class RunMetrics {
+  @pragma('vm:entry-point')
+  RunMetrics._(
+    this.top,
+    this.ascent,
+    this.descent,
+    this.bottom,
+    this.leading,
+    this.avgCharWidth,
+    this.maxCharWidth,
+    this.xMin,
+    this.xMax,
+    this.xHeight,
+    this.capHeight,
+    this.underlineThickness,
+    this.underlinePosition,
+    this.strikeoutThickness,
+    this.strikeoutPosition,
+  );
+
+  // TODO(garyq): Include the TextStyle here too.
+
+  /// Extent above baseline.
+  final double top;
+  /// Distance to reserve above baseline.
+  ///
+  /// Since this is the raw metric from the font, [ascent] is usually negative.
```

Comment:
Is this only true for ascent? Or also for some of the other metrics here?

Reply:
Let me double check, but it may also apply to `top`, although we dont track top separately so an `abs` is never applied to it.